File: frameworks/dotnet/aspnet-mvc/Implementation/Controllers/StreamController.cs
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

namespace Implementation.Controllers;

/// <summary>
/// stream: items.medium's rows written one per line and flushed as each is written. MVC has no
/// result that writes NDJSON, so the action writes to the response itself. The length is never
/// known, so the answer goes out chunked.
/// </summary>
[ApiController]
public sealed class StreamController(Payloads payloads) : ControllerBase
{
    private static readonly byte[] NewLine = "\n"u8.ToArray();

    [HttpGet("/stream/items")]
    public async Task Rows()
    {
        Response.ContentType = "application/x-ndjson";
        foreach (Item row in payloads.Medium.Items)
        {
            await JsonSerializer.SerializeAsync(Response.Body, row, JsonContext.Default.Item);
            await Response.Body.WriteAsync(NewLine);
            await Response.Body.FlushAsync();
        }
    }
}

File: frameworks/node/nestjs/Implementation/stream/stream.controller.ts
import { Readable } from 'node:stream';

import { Controller, Get, Inject, StreamableFile } from '@nestjs/common';

import { PAYLOADS, type Payloads } from '../payloads.js';

/**
 * stream: items.medium's rows written one per line, each as it is produced. StreamableFile, Nest's
 * streamed answer, pipes the lines to the response with no length, so it goes out chunked.
 */
@Controller('stream')
export class StreamController {
  constructor(@Inject(PAYLOADS) private readonly payloads: Payloads) {}

  // rb:handler stream.ndjson
  @Get('items')
  lines(): StreamableFile {
    const rows = this.payloads.medium.items;
    return new StreamableFile(Readable.from(rows.map((row) => `${JSON.stringify(row)}\n`)), { type: 'application/x-ndjson' });
  }
}
